File: src/main/scala/nl.knaw.dans.easy.dd2d/DepositMigrationTask.scala
/*
 * Copyright (C) 2022 DANS - Data Archiving and Networked Services (info@dans.knaw.nl)
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
package nl.knaw.dans.easy.dd2d

import better.files.File
import nl.knaw.dans.easy.dd2d.dansbag.InformationPackageType.InformationPackageType
import nl.knaw.dans.easy.dd2d.dansbag.{ DansBagValidator, InformationPackageType }
import nl.knaw.dans.easy.dd2d.mapping.Amd
import nl.knaw.dans.lib.dataverse.DataverseClient
import nl.knaw.dans.lib.dataverse.model.dataset.Dataset

import java.net.URI
import java.util.regex.Pattern
import scala.language.postfixOps
import scala.util.{ Failure, Success, Try }
import scala.xml.{ Elem, Node }

class DepositMigrationTask(deposit: Deposit,
                           optFileExclusionPattern: Option[Pattern],
                           zipFileHandler: ZipFileHandler,
                           depositorRole: String,
                           deduplicate: Boolean,
                           activeMetadataBlocks: List[String],
                           optDansBagValidator: Option[DansBagValidator],
                           dataverseClient: DataverseClient,
                           publishAwaitUnlockMaxNumberOfRetries: Int,
                           publishAwaitUnlockMillisecondsBetweenRetries: Int,
                           narcisClassification: Elem,
                           iso1ToDataverseLanguage: Map[String, String],
                           iso2ToDataverseLanguage: Map[String, String],
                           variantToLicense: Map[String, String],
                           supportedLicenses: List[URI],
                           reportIdToTerm: Map[String, String],
                           outboxDir: File)
  extends DepositIngestTask(deposit,
    optFileExclusionPattern,
    zipFileHandler,
    depositorRole,
    deduplicate,
    activeMetadataBlocks,
    optDansBagValidator,
    dataverseClient,
    publishAwaitUnlockMaxNumberOfRetries,
    publishAwaitUnlockMillisecondsBetweenRetries,
    narcisClassification,
    iso1ToDataverseLanguage,
    iso2ToDataverseLanguage,
    variantToLicense,
    supportedLicenses,
    reportIdToTerm,
    outboxDir) {
  override protected val informationPackageType: InformationPackageType = InformationPackageType.MIGRATION

  override protected def checkDepositType(): Try[Unit] = {
    for {
      _ <- if (deposit.doi.isEmpty) Failure(new IllegalArgumentException("Deposit for migrated dataset MUST have deposit property identifier.doi set"))
           else Success(())
      _ <- deposit.vaultMetadata.checkMinimumFieldsForImport()
    } yield ()
  }

  override def newDatasetUpdater(dataverseDataset: Dataset): DatasetUpdater = {
    new DatasetUpdater(deposit, optFileExclusionPattern, zipFileHandler, isMigration = true, dataverseDataset.getDatasetVersion.getMetadataBlocks, variantToLicense, supportedLicenses, dataverseClient)
  }

  override def newDatasetCreator(dataverseDataset: Dataset, depositorRole: String): DatasetCreator = {
    new DatasetCreator(deposit, optFileExclusionPattern, zipFileHandler, depositorRole, isMigration = true, dataverseDataset, variantToLicense, supportedLicenses, dataverseClient)
  }

  override protected def checkPersonalDataPresent(optAgreements: Option[Node]): Try[Unit] = {
    if (optAgreements.isEmpty) Failure(RejectedDepositException(deposit, "Migration deposit MUST have an agreements.xml"))
    else Success(())
  }

  override protected def getDateOfDeposit: Try[Option[String]] = {
    for {
      optAmd <- deposit.tryOptAmd
      optDate = optAmd.flatMap(Amd toDateOfDeposit)
    } yield optDate
  }

  override protected def publishDataset(persistentId: String): Try[Unit] = {
    trace(persistentId)
    for {
      optAmd <- deposit.tryOptAmd
      amd = optAmd.getOrElse(throw new Exception(s"no AMD found for $persistentId"))
      optPublicationDate <- getJsonLdPublicationdate(amd)
      publicationDate = optPublicationDate.getOrElse(throw new IllegalArgumentException(s"no publication date found in AMD for $persistentId"))
      _ <- Try(dataverseClient.dataset(persistentId).releaseMigrated(publicationDate, true))
      _ <- Try(dataverseClient.dataset(persistentId).awaitUnlock(
        publishAwaitUnlockMaxNumberOfRetries,
        publishAwaitUnlockMillisecondsBetweenRetries))
    } yield ()
  }

  private def getJsonLdPublicationdate(amd: Node): Try[Option[String]] = Try {
    Amd.toPublicationDate(amd)
      .map(d => s"""{"http://schema.org/datePublished": "$d"}""")
  }

  override protected def postPublication(persistentId: String): Try[Unit] = {
    trace(persistentId)
    Success(())
  }
}

File: src/main/java/nl/knaw/dans/ingest/core/service/DepositMigrationTask.java
/*
 * Copyright (C) 2022 DANS - Data Archiving and Networked Services (info@dans.knaw.nl)
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
package nl.knaw.dans.ingest.core.service;

import lombok.extern.slf4j.Slf4j;
import nl.knaw.dans.ingest.core.dataverse.DatasetService;
import nl.knaw.dans.ingest.core.deposit.DepositManager;
import nl.knaw.dans.ingest.core.domain.Deposit;
import nl.knaw.dans.ingest.core.domain.DepositLocation;
import nl.knaw.dans.ingest.core.domain.VaultMetadata;
import nl.knaw.dans.ingest.core.exception.FailedDepositException;
import nl.knaw.dans.ingest.core.exception.InvalidDatasetStateException;
import nl.knaw.dans.ingest.core.exception.RejectedDepositException;
import nl.knaw.dans.ingest.core.service.mapper.DepositToDvDatasetMetadataMapper;
import nl.knaw.dans.ingest.core.service.mapper.DepositToDvDatasetMetadataMapperFactory;
import nl.knaw.dans.ingest.core.service.mapper.mapping.Amd;
import nl.knaw.dans.ingest.core.validation.DepositorAuthorizationValidator;
import nl.knaw.dans.lib.dataverse.DataverseException;
import nl.knaw.dans.lib.dataverse.model.dataset.Dataset;
import nl.knaw.dans.validatedansbag.client.api.ValidateCommandDto;
import org.apache.commons.lang3.StringUtils;

import java.io.IOException;
import java.net.URI;
import java.nio.file.Path;
import java.util.ArrayList;
import java.util.List;
import java.util.Optional;
import java.util.regex.Pattern;
import java.util.stream.Collectors;

@Slf4j
public class DepositMigrationTask extends DepositIngestTask {
    public DepositMigrationTask(
        DepositToDvDatasetMetadataMapperFactory datasetMetadataMapperFactory,
        DepositLocation depositLocation,
        String depositorRole,
        Pattern fileExclusionPattern,
        ZipFileHandler zipFileHandler,
        List<URI> supportedLicenses,
        DansBagValidator dansBagValidator,
        Path outboxDir,
        EventWriter eventWriter,
        DepositManager depositManager,
        DatasetService datasetService,
        BlockedTargetService blockedTargetService,
        DepositorAuthorizationValidator depositorAuthorizationValidator,
        String vaultMetadataKey,
        boolean deleteDraftOnFailure
    ) {
        super(
            datasetMetadataMapperFactory, depositLocation, depositorRole, fileExclusionPattern, zipFileHandler, supportedLicenses, dansBagValidator,
            outboxDir, eventWriter, depositManager, datasetService, blockedTargetService, depositorAuthorizationValidator, vaultMetadataKey, deleteDraftOnFailure);
    }

    @Override
    void checkDoiRequirements() {
        if (StringUtils.isEmpty(deposit.getDoi())) {
            throw new IllegalArgumentException("Deposit for migrated dataset MUST have deposit property identifier.doi set");
        }

        validateVaultMetadata(deposit.getVaultMetadata());
    }

    void validateVaultMetadata(VaultMetadata vaultMetadata) {
        var missing = new ArrayList<String>();

        if (StringUtils.isBlank(vaultMetadata.getPid())) {
            missing.add("dataversePid");
        }

        if (StringUtils.isBlank(vaultMetadata.getNbn())) {
            missing.add("dataverseNbn");
        }

        if (!missing.isEmpty()) {
            var msg = String.join(", ", missing);
            throw new RuntimeException(String.format(
                "Not enough Data Vault Metadata for import deposit, missing: %s", msg
            ));
        }
    }

    DepositToDvDatasetMetadataMapper newMapper() {
        return datasetMetadataMapperFactory.createMapper(false, true);
    }

    @Override
    DatasetEditor newDatasetCreator(Dataset dataset, String depositorRole) {
        return newDatasetCreator(dataset, depositorRole, true);
    }

    @Override
    DatasetEditor newDatasetUpdater(Dataset dataset) {
        return newDatasetUpdater(dataset, true, deleteDraftOnFailure);
    }

    @Override
    Optional<String> getDateOfDeposit() {
        return Optional.ofNullable(deposit.getAmd())
            .map(Amd::toDateOfDeposit)
            .flatMap(i -> i);
    }

    @Override
    void publishDataset(String persistentId) throws IOException, DataverseException {
        var amd = deposit.getAmd();

        if (amd == null) {
            throw new RuntimeException(String.format("no AMD found for %s", persistentId));
        }

        var date = Amd.toPublicationDate(amd);

        if (date.isEmpty()) {
            throw new IllegalArgumentException(String.format("no publication date found in AMD for %s", persistentId));
        }

        datasetService.releaseMigrated(persistentId, date.get());
    }

    void postPublication(String persistentId) throws IOException, DataverseException, InterruptedException {
        try {
            datasetService.waitForState(persistentId, "RELEASED");
            // Do NOT save persistent identifiers, as they were provided in the deposit
        } catch (InvalidDatasetStateException e) {
            throw new FailedDepositException(deposit, e.getMessage());
        }
    }

    void validateDeposit() {
        var result = dansBagValidator.validateBag(
            deposit.getBagDir(), ValidateCommandDto.PackageTypeEnum.MIGRATION);

        if (!result.getIsCompliant()) {
            var violations = result.getRuleViolations().stream()
                .map(r -> String.format("- [%s] %s", r.getRule(), r.getViolation()))
                .collect(Collectors.joining("\n"));

            throw new RejectedDepositException(deposit, String.format(
                "Bag was not valid according to Profile Version %s. Violations: %s",
                result.getProfileVersion(), violations)
            );
        }
    }
}

